src/control/destination/background/mod.rs

```diff
@@ -194,12 +198,20 @@ where
                     let config = &self.config;
                     let dsts = &mut self.dsts;
 
+                    // If the requested authority currently needs more
+                    // query capacity to query the destination service, go
+                    // ahead and try to free up capacity (by dropping any
+                    // inactive DestinationSets).
+                    if dsts.needs_query_for(&resolve.authority) {
```

Comment:
Logically, it seems like we should do this only inside of the pattern match, after we know the entry for the given authority is occupied. However, `dsts.destinations.entry(...)` borrows `dsts` mutably, so we can't call `dsts.retain_active()` there. So, we have to do this here, instead.

The overhead of just `retain_active`-ing every time is hopefully reduced by pre-checking that it would actually be _necessary_.